truffle/src/com.oracle.truffle.runtime/src/com/oracle/truffle/runtime/OptimizedTruffleRuntime.java

```diff
@@ -912,10 +913,45 @@ private void notifyCompilationFailure(OptimizedCallTarget callTarget, Throwable
     protected void onEngineCreated(EngineData engine) {
     }
 
+    private long stoppedCompilationTime = 0;
+    private boolean logShutdownCompilations = true;
+
     @SuppressWarnings("try")
     public CompilationTask submitForCompilation(OptimizedCallTarget optimizedCallTarget, boolean lastTierCompilation) {
-        Priority priority = new Priority(optimizedCallTarget.getCallAndLoopCount(), lastTierCompilation ? Priority.Tier.LAST : Priority.Tier.FIRST);
-        return getCompileQueue().submitCompilation(priority, optimizedCallTarget);
+        BackgroundCompileQueue queue = getCompileQueue();
+        CompilationActivityMode compilationActivityMode = getCompilationActivityMode();
+        if (compilationActivityMode == CompilationActivityMode.RUN_COMPILATION ||
+                (stoppedCompilationTime != 0 && System.currentTimeMillis() - stoppedCompilationTime > optimizedCallTarget.getOptionValue(StoppedCompilationRetryDelay))) {
+            stoppedCompilationTime = 0;
+            Priority priority = new Priority(optimizedCallTarget.getCallAndLoopCount(), lastTierCompilation ? Priority.Tier.LAST : Priority.Tier.FIRST);
+            return queue.submitCompilation(priority, optimizedCallTarget);
+        } else if (compilationActivityMode == CompilationActivityMode.STOP_COMPILATION) {
+            if (stoppedCompilationTime == 0) {
+                stoppedCompilationTime = System.currentTimeMillis();
+            }
+            // Flush the compilations queue. There's still a chance that compilation will be re-enabled
+            // eventually, if the hosts code cache can be cleaned up.
+            for (OptimizedCallTarget target : queue.getQueuedTargets(optimizedCallTarget.engine)) {
+                target.cancelCompilation("Compilation temporary disabled due to full code cache.");
+            }
+        } else {
+            // Compilation was shut down permanently because the hosts code cache ran full and
+            // the host was configured without support for code cache sweeping.
+            assert compilationActivityMode == CompilationActivityMode.SHUTDOWN_COMPILATION;
+            TruffleLogger logger = optimizedCallTarget.engine.getLogger("engine");
+            // The logger can be null if the engine is closed.
+            if (logger != null && logShutdownCompilations) {
+                logShutdownCompilations = false;
+                logger.log(Level.WARNING, "Truffle host compilations permanently disabled because of full code cache. "  +
+                    "Increase the code cache size using '-XX:ReservedCodeCacheSize=' and/or run with '-XX:+UseCodeCacheFlushing -XX:+MethodFlushing'.");
+            }
+            try {
+                queue.shutdownAndAwaitTermination(100 /* milliseconds */);
```

Comment:
Can you describe a bit what you are trying to achieve here? why would it help to block the entire interpreter thread if this happens?

Reply:
The fastest way to disable a compilation is using OptimizedCallTarget.compilationFailed.
Every time we see shutdown with a call target we could disable compilation.
Best idea I have.
I put up an example what i would do in another comment (no idea how to link draft comments here)